compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1297,115 +1297,240 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
-              val lhsVal = lhsCore.denot.suchThat(!_.is(Method))
-              val lhsSym = lhsVal.symbol
-              if canAssign(lhsSym) then
-                rememberNonLocalAssignToPrivate(lhsSym)
-                // lhsBounds: (T .. Any) as seen from lhs prefix, where T is the type of lhsSym
-                // This ensures we do the as-seen-from on T with variance -1. Test case neg/i2928.scala
-                val lhsBounds =
-                  TypeBounds.lower(lhsSym.info).asSeenFrom(ref.prefix, lhsSym.owner)
-                assignType(cpy.Assign(tree)(lhs1, typed(tree.rhs, lhsBounds.loBound)))
-                  .computeAssignNullable()
-              else
-                val pre = ref.prefix
-                val setterName = ref.name.setterName
-                val setter = pre.member(setterName)
-                lhsCore match {
-                  case lhsCore: RefTree if setter.exists =>
-                    val setterTypeRaw = pre.select(setterName, setter)
-                    val setterType = ensureAccessible(setterTypeRaw, isSuperSelection(lhsCore), tree.srcPos)
-                    val lhs2 = untpd.rename(lhsCore, setterName).withType(setterType)
-                    typedUnadapted(untpd.Apply(untpd.TypedSplice(lhs2), tree.rhs :: Nil), WildcardType, locked)
-                  case _ =>
-                    reassignmentToVal
+  /** Returns a builder for making trees representing assignments to `lhs`, which isn't a term or
+    * type application.
+    */
+  def formPartialAssignmentToNonApply(
+      lhs: untpd.Tree, isSingleAssignment: Boolean
+  )(using Context): PartialAssignment[LValue] =
+    val locked = ctx.typerState.ownedVars
+    val core = typedUnadapted(lhs, LhsProto, locked)
+    def adapted = adapt(core, LhsProto, locked)
+
+    /** Returns a builder reporting that the left-hand side is not reassignable. */
+    def reassignmentToVal(): PartialAssignment[SimpleLValue] =
+      PartialAssignment(SimpleLValue(core)) { (l, r) =>
+        val target = l.expression
+        report.error(ReassignmentToVal(target.symbol.name), target.srcPos)
+        untpd.TypedSplice(tpd.Assign(target, typed(r, adapted.tpe.widen)))
+      }
+
+    /** Returns `true` if `s` is assignable. */
+    def canAssign(s: Symbol) =
+      s.is(Mutable, butNot = Accessor) ||
+      ctx.owner.isPrimaryConstructor && !s.is(Method) && s.maybeOwner == ctx.owner.owner ||
+      // allow assignments from the primary constructor to class fields
+      ctx.owner.name.is(TraitSetterName) || ctx.owner.isStaticConstructor
+
+    /** Marks private variables that are assigned with a prefix other than the `this` type of their
+     *  owner with a `annotation.internal.AssignedNonLocally` annotation. The annotation influences
+     *  the variance check for these variables, which is done at PostTyper. It will be removed
+     *  after the variance check.
+     */
+    def rememberNonLocalAssignToPrivate(s: Symbol) = adapted match
```

Comment:
More in this method. It seems like it could be unchanged modulo indentation compared to the code before.